text/0000-ds-dlopen-error-handling.md

```diff
@@ -0,0 +1,204 @@
+auto svcRef = framework.GetBundleContext().GetServiceReference<Foo>();
+
+// This will throw a SharedLibraryException if the shared library failed to load.
+// either allow the throw to propagate or try/catch and do something
+try {
+  auto svc = framework.GetBundleContext().GetService(svcRef);
+} catch(const cppmicroservices::SharedLibraryException& ex) {
+  // do something with ex...
+}
+
+if(nullptr == svc) {
+  // do something knowing that the reason this is nullptr is due to a recoverable failure.
+}
+```
+
+
+
+#### CppMicroServices implementation changes
+
+- `ServiceReferenceBasePrivate::GetServiceFromFactory` needs to explicitly catch `SharedLibraryException` before catching `std::exception`, log a `FrameworkEvent` and rethrow.
+- `Bundle::Start()` needs to throw `SharedLibraryException` if `dlopen/LoadLibrary` failed.
+
+
+
+#### DS implementation changes
+
+- Declarative Services needs to throw a `SharedLibraryException` in two cases; 
+  - When the service component has `immediate` set to true.
+    - `Bundle::Start()` will throw a `SharedLibraryException`. 
+  - When the service component has `immediate` set to false. 
+    -  `BundleContext::GetService()` will throw a `SharedLibraryException`.
+    - Declarative Services has try/catch blocks around the operations to enable a service component. These try/catch blocks will have to be modified to catch a `SharedLibraryException` and rethrow it. The try/catch block starts in `SCRActivator::CreateExtension` with more try/catch blocks contained within the call chain of this method, ultimately ending in `BundleLoader::GetComponentCreatorDeletors` where the `dlopen/LoadLibrary` happens. 
+
+
+
+## How we teach this
+
+> What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing CppMicroServices patterns, or as a
+wholly new one?
+
+> Would the acceptance of this proposal mean the CppMicroServices guides must be
+re-organized or altered? Does it change how CppMicroServices is taught to new users
+at any level?
+
+> How should this feature be introduced and taught to existing CppMicroServices
+users?
+
+- Requires updates to the doxygen documentation for  `BundleContext::GetService` and `Bundle::Start()`describing the new exception type which can be thrown. 
+
+
+
+## Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching CppMicroServices,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+> There are tradeoffs to choosing any path, please attempt to identify them here.
+
+- Not OSGi compliant. The relevant OSGi text can be found under ServiceFactory::GetService where it describes how errors coming from service factories (which DS uses to do its job) should be handled:
```

Comment:
If Java OSGi throws "class not found" exceptions when the class loader can't find a class, then is throwing a "shared library loading failed" exception the equivalent and therefore is this not a violation of the OSGi spec?

Reply:
@kevinleeMW please remove this as a drawback. I think it's reasonable to equate this 'SharedLibraryException` exception with the Java "class not found" exceptions. Also the implementation does log the `SharedLibraryException` as a `FrameworkEvent` before re-throwing it. thanks.